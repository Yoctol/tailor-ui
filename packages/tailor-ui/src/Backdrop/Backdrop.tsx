import React, { FunctionComponent } from 'react';
import styled from 'styled-components';
import { animated, config, useTransition } from 'react-spring';

import Portal from '../Portal';

const StyledBackdrop = styled.div`
  position: fixed;
  z-index: -1;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: rgba(0, 0, 0, 0.65);
`;

const AnimatedStyledBackdrop = animated(StyledBackdrop);

export interface BackdropProps {
  visible: boolean;
  [key: string]: any;
}

const Backdrop: FunctionComponent<BackdropProps> = ({
  visible,
  ...otherProps
}) => {
  const transition = useTransition(visible, null, {
    from: {
      opacity: 0,
    },
    enter: {
      opacity: 1,
    },
    leave: {
      opacity: 0,
      pointerEvents: 'none',
    },
    config: config.stiff,
  });

  return (
    // FIXME: react type
    <>
      {transition.map(
        ({ item, key, props }) =>
          item && (
            <Portal key={key}>
              <AnimatedStyledBackdrop style={props} {...otherProps} />
            </Portal>
          )
      )}
    </>
  );
};

export default Backdrop;
