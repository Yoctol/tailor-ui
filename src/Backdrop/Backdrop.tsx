import React, { FC, HTMLAttributes } from 'react';
import styled from 'styled-components';
import { animated, config, useTransition } from 'react-spring';

import { Portal } from '../Portal';
import { StackingOrder } from '../constants';

const StyledBackdrop = styled.div`
  position: fixed;
  z-index: -1;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: rgba(0, 0, 0, 0.65);
`;

const AnimatedStyledBackdrop = animated(StyledBackdrop);

export interface BackdropProps extends HTMLAttributes<HTMLDivElement> {
  visible: boolean;
  zIndex?: number;
}

const Backdrop: FC<BackdropProps> = ({
  visible,
  zIndex = StackingOrder.OVERLAY,
  ...otherProps
}) => {
  const transition = useTransition(visible, null, {
    from: {
      opacity: 0,
    },
    enter: {
      opacity: 1,
    },
    leave: {
      opacity: 0,
      pointerEvents: 'none',
    },
    config: config.stiff,
  });

  return (
    // FIXME: react type
    <>
      {transition.map(
        ({ item, key, props }) =>
          item && (
            <Portal key={key} defaultOrder={zIndex}>
              <AnimatedStyledBackdrop style={props} {...otherProps} />
            </Portal>
          )
      )}
    </>
  );
};

export { Backdrop };
