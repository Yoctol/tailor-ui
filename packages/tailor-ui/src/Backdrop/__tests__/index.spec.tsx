import React from 'react';

import { render } from 'test/test-utils';

import Backdrop from '../Backdrop';

describe('Backdrop', () => {
  it('should render backdrop correctly', async () => {
    const { baseElement } = render(<Backdrop visible />);

    expect(baseElement).toMatchSnapshot();
  });

  it('should not render backdrop', async () => {
    const { baseElement } = render(<Backdrop visible={false} />);

    expect(baseElement).toMatchSnapshot();
  });
});
