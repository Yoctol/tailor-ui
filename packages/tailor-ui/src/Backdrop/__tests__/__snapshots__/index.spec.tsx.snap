// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backdrop should not render backdrop 1`] = `null`;

exports[`Backdrop should render backdrop correctly 1`] = `
.c0 {
  position: fixed;
  z-index: -1;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: rgba(0,0,0,0.65);
}

<div
  class="c0"
  style="opacity: 1;"
/>
`;
